### PR TITLE
Add handling for state transfer blockchain

### PR DESCRIPTION
**Pull request by cloudnoize**
Defining a class (```detail::Blockchain::StateTransfer```) that manages the interaction with the state transfer blockchain.
Migrating the caching logic of ```last_reachable_block_id``` and ```last_block_id``` i.e. last state transfer block id.

Implementing the ```add``` and ```get``` operations on raw blocks via the kv_blockchain, which may result in linking
blocks to the main blockchain by removing them atomically from the st blockchain.

Changing the raw block CMF structs in order to be compatible with the kv_blockchain ```addBlock``` and enabling ability to ```move``` them.

not related - renaming of metrics.hpp -> s3_metrics.hpp in order to eliminate compilation error on macos filesystem.

### CODE REVIEW

File: kvbc/cmf/categorized_kvbc_msgs.cmf [hunk 1]
```diff
@@ -73,28 +73,18 @@ Msg BlockData 2001 {
...
 Msg RawBlockData 2005 {
     fixedlist uint8 32 parent_digest
+    
     map string oneof{
-        RawBlockMerkleUpdates
-        RawBlockKeyValueUpdates
-        RawBlockImmutableUpdates
+        MerkleUpdatesData
+        KeyValueUpdatesData
+	ImmutableUpdatesData
```
Comment (dartdart26) [inline kvbc/cmf/categorized_kvbc_msgs.cmf:83]: Nit: Indent.

File: kvbc/src/categorization/kv_blockchain.cpp [hunk 2]
```diff
@@ -240,4 +247,78 @@ ImmutableUpdatesInfo KeyValueBlockchain::handleCategoryUpdates(
...
+  for (auto i = block_id; i <= last_block_id.value(); ++i) {
+    auto raw_block = state_transfer_block_chain_.getRawBlock(i);
+    if (!raw_block) {
+      return;
+    }
+    writeSTLinkTransaction(block_id, *raw_block);
+  }
+  // Linking has fully completed and we should not have any more ST temporary blocks left. Therefore, make sure we don't
+  // have any value for the latest ST temporary block ID cache.
+  state_transfer_block_chain_.resetChain();
```
Comment (dartdart26) [inline kvbc/src/categorization/kv_blockchain.cpp:310]: Great idea to introduce the ST chain as a separate object!

File: kvbc/src/categorization/kv_blockchain.cpp [hunk 3]
```diff
@@ -240,4 +247,78 @@ ImmutableUpdatesInfo KeyValueBlockchain::handleCategoryUpdates(
...
+}
+
+// Atmoic delete from state transfer and add to blockchain
```
Comment (dartdart26) [inline kvbc/src/categorization/kv_blockchain.cpp:313]: Typo: Atomic.

File: kvbc/src/categorization/kv_blockchain.cpp [hunk 4]
```diff
@@ -240,4 +247,78 @@ ImmutableUpdatesInfo KeyValueBlockchain::handleCategoryUpdates(
...
+void KeyValueBlockchain::writeSTLinkTransaction(const BlockId block_id, RawBlock& block) {
+  auto write_batch = native_client_->getBatch();
+  state_transfer_block_chain_.deleteBlock(block_id, write_batch);
+  auto new_block_id = addBlock(std::move(block.data.category_updates), write_batch);
+  native_client_->write(std::move(write_batch));
+
+  block_chain_.setAddedBlockId(new_block_id);
+  state_transfer_block_chain_.updateLastIdAfterDeletion(block_id);
```
Comment (dartdart26) [inline kvbc/src/categorization/kv_blockchain.cpp:321]: I am not sure we need to call `updateLastIdAfterDeletion()` here as we don't change the last ID of the ST chain. Doing so works, but it does a load that in turn seeks and, if I am not missing something, is unnecessary. Please correct me if I am wrong.

File: kvbc/test/categorization/blockchain_test.cpp [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+// "License").  You may not use this product except in compliance with the
+// Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+/**
+ * The following test suite tests ordering of KeyValuePairs
```
Comment (dartdart26) [inline kvbc/test/categorization/blockchain_test.cpp:15]: Nit: I think that might be an incorrect comment.

File: kvbc/test/categorization/blockchain_test.cpp [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+using namespace concord::kvbc;
+
+namespace {
+
+class categorized_kvbc : public ::testing::Test {
+  void SetUp() override {
+    cleanup();
+    db = TestRocksDb::createNative();
+    db->createColumnFamily(BLOCKS_CF);
+    db->createColumnFamily(ST_CHAIN_CF);
```
Comment (dartdart26) [inline kvbc/test/categorization/blockchain_test.cpp:41]: Maybe the KV blockchain should create the column families (if not existing) on construction?

File: kvbc/test/categorization/blockchain_test.cpp [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+
+  // delete non-existing block
+  {
+    auto wb = db->getBatch();
+    block_chain.deleteBlock(200, wb);
+    db->write(std::move(wb));
+    ASSERT_EQ(block_chain.loadGenesisBlockId().value(), 1);
+  }
+
+  // delete last and load
```
Comment (dartdart26) [inline kvbc/test/categorization/blockchain_test.cpp:114]: Nit: "delete the genesis and load"

File: kvbc/test/categorization/kv_blockchain_test.cpp [hunk 8]
```diff
@@ -229,6 +229,255 @@ TEST_F(categorized_kvbc, delete_block) {
...
+  {
+    categorization::RawBlock rb;
+    block_chain.addRawBlock(rb, 6);
+    ASSERT_TRUE(block_chain.getLastStatetransferBlockId().has_value());
+    ASSERT_EQ(block_chain.getLastStatetransferBlockId().value(), 6);
+
+    ASSERT_EQ(block_chain.getLastReachableBlockId(), 2);
+    ASSERT_EQ(block_chain_imp.loadLastReachableBlockId().value(), 2);
+  }
+  // Link the chains
```
Comment (dartdart26) [inline kvbc/test/categorization/kv_blockchain_test.cpp:466]: Nice test! Do we need to also verify the contents (via `getBlock()`) of a raw block after it has been linked?

**Review by dartdart26 — APPROVED** [review]
Looks great, thanks @cloudnoize ! Very clear and nice separation!

The only thing that we should maybe fix is the removal of the seek when linking the two chains.

File: kvbc/src/categorization/kv_blockchain.cpp (hunk 4 above)
Comment (cloudnoize) [reply kvbc/src/categorization/kv_blockchain.cpp:321]: You're right, in this context is redundant, I'll remove it

(Though it won't actually do anything due to the condition at the start 
```if (!last_block_id_.has_value() || last_block_id_.value() != id) {
        return;
   }```)

File: kvbc/test/categorization/kv_blockchain_test.cpp (hunk 8 above)
Comment (cloudnoize) [reply kvbc/test/categorization/kv_blockchain_test.cpp:466]: yeah, but I'll do it after wiring the categories to actually produce the updates.

File: kvbc/test/categorization/blockchain_test.cpp (hunk 6 above)
Comment (cloudnoize) [reply kvbc/test/categorization/blockchain_test.cpp:41]: Probably, the constructor now is trivial, I'll start handling it in the following PR I believe.

**Pull request merged**